### PR TITLE
Refactoring code

### PR DESCRIPTION
**Pull request by rachitnigam**
* Moved the code for project into a different file.
* Warning added when undefined is returned on a object lookup.
* Better error messages.

### CODE REVIEW

File: js/src/main/scala/FunctionLib.scala [hunk 1]
```diff
@@ -44,45 +44,29 @@ object FunctionLib{
...
   val mul = jsBinArith(_ * _) _
   val div = jsBinArith(_ / _) _
   val mod = jsBinArith(_ % _) _
 
   val funs: Map[String, List[Value] => Value] =
     Map(
       "prim_add" -> add,
-      "sub" -> sub,
-      "inc" -> inc,
-      "dec" -> dec,
```
Comment (arjunguha) [inline js/src/main/scala/FunctionLib.scala]: almost overkill. but this is fine

Comment (rachitnigam) [inline js/src/main/scala/FunctionLib.scala]: I thought you liked code deletion!

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 2]
```diff
@@ -326,20 +246,24 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
...
-      store.deref(loc) match {
-        case VJSObj(obj) => obj.asInstanceOf[js.Dynamic]
-        // NOTE: We are passing the underlying array, which may be updated in JavaScript. Ensure that this
-        // is only passed to functions that don't calculate with the actual array elements.
-        case VArray(items, _) => items.asInstanceOf[js.Dynamic]
-        // NOTE: we are assuming that the first-order function doesn't examine the fields in the array.
-        // E.g., this definition makes since for the .push and .unslice methods on arrays.
-        case VObj(_) => loc.asInstanceOf[js.Dynamic]
-        case invalid => throw Unexpected(s"Cannot pass $invalid (in the store) to a first-order JS function")
+    case (loc: VHeapRef) => {
```
Comment (arjunguha) [inline js/src/main/scala/JavaScriptCEKLike.scala]: What happened here? Why did you reorder these cases? (This is just a permutation, right?)

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 3]
```diff
@@ -374,7 +297,9 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
         actuals match {
           case thisArg :: otherArgs =>
             VConfig(shallowFromJs(f(shallowToJs(thisArg), js.Array(otherArgs.map(shallowToJs): _*))), k)
-          case _ => throw Unexpected("no this-argument when applying a first-order builtin function")
+          case _ => throw Unexpected(
```
Comment (arjunguha) [inline js/src/main/scala/JavaScriptCEKLike.scala]: What does this achieve? 80 characters?

Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: Yes

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 4]
```diff
@@ -397,45 +324,45 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
...
-            val env2 = if (usesArguments == false) {
-              Map.empty
-            }
-            else {
-              // We drop the "this"-argument
-              Map(("arguments" -> store.alloc(store.alloc(VArray(js.Array(actuals.drop(1): _*), VObj(Map.empty))))))
-            }
-            EConfig(body, env ++ env0 ++ env2, k)
+    case VJSObj(jsFun) => VConfig(jsApply(jsFun, actuals), k)
+    case obj@VObj(exprMap) => exprMap.get("*code") match {
```
Comment (arjunguha) [inline js/src/main/scala/JavaScriptCEKLike.scala]: This refactoring seems wrong.

The real problem is that VObj is not a value. I agree that the original code had some duplication, but the refactored code actually assumes that VObj is a value. It isn't any better and is probably a little worse, IMO.

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 5]
```diff
@@ -625,16 +555,26 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
                 toInt(keyStr) match {
                   // The apply method on js.Array blows up badly if the element is not defined.
                   case Some(a) if (a >= 0) => items.asInstanceOf[js.Dynamic].selectDynamic(keyStr).asInstanceOf[Any] match {
-                    case () => VConfig(VConst(CUndefined()), k)
+                    case () => {
+                      logger.warn(s"Returning undefined for index $a for the array $obj")
```
Comment (arjunguha) [inline js/src/main/scala/JavaScriptCEKLike.scala]: Why the warning?

**Comment by arjunguha** [comment]
As you can see, I'm happy with some changes but not others. In the interest of getting the A+ changes in, why don't you split this into two pull requests (1) the consolidated operators and the projectExpr extraction and (2) the changes to applyFunction. I'll merge (1). I want to discuss (2) a little more before we do.

File: js/src/main/scala/JavaScriptCEKLike.scala (hunk 2 above)
Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: Trying to remove the `case _ if _ =>` patterns.

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 6]
```diff
@@ -387,7 +312,9 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
     // This case allows the AST produced by desugaring to insert direct calls to builtin functions.
     case VBuiltinFunction(name) => {
       if (actuals.forall(v => v.isInstanceOf[VFacet] == false)) {
-        val f = FunctionLib.builtins.getOrElse(name, throw Unexpected(s"$name is not a builtin function"))
+        val f = FunctionLib.builtins.getOrElse(
+          name, throw Unexpected(s"$name is not a builtin function")
+        )
```
Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: 80 chars.

File: js/src/main/scala/JavaScriptCEKLike.scala (hunk 5 above)
Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: It is useful in tracking down non-local undefined bugs.

File: js/src/main/scala/JavaScriptCEKLike.scala (hunk 4 above)
Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: I was unsure about the refactoring, but the code hit a case where a `VObj(VClosure(_))` occurred in function position. Apart from that the `VFacet(_)` case also needed to be duplicated out the body.

I don't entirely understand the statement `VObj` is not a value. Why is this the case?

File: js/src/main/scala/JavaScriptCEKLike.scala (hunk 2 above)
Comment (arjunguha) [inline js/src/main/scala/JavaScriptCEKLike.scala]: Where is the `case _ if _ =>` that you removed?

File: js/src/main/scala/JavaScriptCEKLike.scala [hunk 7]
```diff
@@ -616,7 +543,10 @@ trait JavaScriptCEKLike extends slogging.LazyLogging  {
               case o@VObj(fields) => fields.get(keyStr) match {
                 case Some(x) => VConfig(x, k)
                 case None => fields.get("__proto__") match {
-                  case Some(VConst(CNull())) => VConfig(VConst(CUndefined()), k)
+                  case Some(VConst(CNull())) => {
+                    logger.warn(s"Returning undefined for $keyStr since $obj doesn't contain the field and doesn't have a __proto__")
```
Comment (arjunguha) [inline js/src/main/scala/JavaScriptCEKLike.scala]: yeah, its useful until we are 100% confident in Fission

File: js/src/main/scala/JavaScriptCEKLike.scala (hunk 2 above)
Comment (rachitnigam) [inline js/src/main/scala/JavaScriptCEKLike.scala]: `case (loc : VHeapRef) if loc.cell.isLocal == false => {`

**Pull request closed without merging**